tests/Microsoft.Data.Analysis.Tests/DataFrameTests.cs

```diff
@@ -1947,84 +1947,170 @@ public void TestMutationOnRows()
+            Assert.Equal(16, ret.Rows.Count);
+
+            Assert.Equal(16, ret["Column1"].Length);
+            Assert.Equal(16, ret["Column2"].Length);
+            Assert.Equal(2, ret.Columns[0].NullCount);
+            Assert.Equal(5, ret.Columns[1].NullCount);
 
-            df.Append(new Dictionary<string, object> { { "Column2", false } });
+            df.Append(new Dictionary<string, object> { { "Column2", false } }, inPlace: true);
             Assert.Equal(16, df.Rows.Count);
             Assert.Equal(16, df["Column1"].Length);
             Assert.Equal(16, df["Column2"].Length);
             Assert.Equal(3, df.Columns[0].NullCount);
             Assert.Equal(4, df.Columns[1].NullCount);
-
-            df.Append((IEnumerable<object>)null);
+            ret = df.Append(new Dictionary<string, object> { { "Column2", false } }, inPlace: false);
+            Assert.Equal(17, ret.Rows.Count);
+            Assert.Equal(17, ret["Column1"].Length);
+            Assert.Equal(17, ret["Column2"].Length);
+            Assert.Equal(4, ret.Columns[0].NullCount);
+            Assert.Equal(4, ret.Columns[1].NullCount);
+
+            df.Append((IEnumerable<object>)null, inPlace: true);
             Assert.Equal(17, df.Rows.Count);
             Assert.Equal(17, df["Column1"].Length);
             Assert.Equal(17, df["Column2"].Length);
             Assert.Equal(4, df.Columns[0].NullCount);
             Assert.Equal(5, df.Columns[1].NullCount);
+            ret = df.Append((IEnumerable<object>)null, inPlace: false);
+            Assert.Equal(18, ret.Rows.Count);
+            Assert.Equal(18, ret["Column1"].Length);
+            Assert.Equal(18, ret["Column2"].Length);
+            Assert.Equal(5, ret.Columns[0].NullCount);
+            Assert.Equal(6, ret.Columns[1].NullCount);
 
             // DataFrame must remain usable even if Append throws
-            Assert.Throws<FormatException>(() => df.Append(new List<object> { 5, "str" }));
-            Assert.Throws<FormatException>(() => df.Append(new Dictionary<string, object> { { "Column2", "str" } }));
-            Assert.Throws<ArgumentException>(() => df.Append(new List<object> { 5, true, true }));
+            Assert.Throws<FormatException>(() => df.Append(new List<object> { 5, "str" }, inPlace: true));
+            Assert.Throws<FormatException>(() => df.Append(new Dictionary<string, object> { { "Column2", "str" } }, inPlace: true));
+            Assert.Throws<ArgumentException>(() => df.Append(new List<object> { 5, true, true }, inPlace: true));
 
-            df.Append();
+            df.Append(inPlace: true);
             Assert.Equal(18, df.Rows.Count);
             Assert.Equal(18, df["Column1"].Length);
             Assert.Equal(18, df["Column2"].Length);
             Assert.Equal(5, df.Columns[0].NullCount);
             Assert.Equal(6, df.Columns[1].NullCount);
+            ret = df.Append(inPlace: false);
+            Assert.Equal(19, ret.Rows.Count);
```

Comment:
Why does `inPlace: true` return 18 rows, but `inPlace: false` return 19 rows?

Reply:
Ah, you're looking at the wrong dataframe. Maybe I set it up badly. After line 2081, `df` has 18 rows. After 2087, `df` still has 18 rows, but `ret` has a new row appended to it. Hence the `19`

Reply:
OK, maybe add an assert here that `df` isn't changed by calling `df.Append(inPlace:false)`.

---
In reply to: [371531652](https://github.com/dotnet/corefxlab/pull/2823#discussion_r371531652) [](ancestors = 371531652)